app/src/main/java/org/dslul/openboard/inputmethod/latin/settings/UserDictionaryListFragment.java

```diff
@@ -0,0 +1,275 @@
+        }
+
+        // Add the main language selected in the "Language and Layouts" setting except "No language"
+        for (InputMethodSubtype mainSubtype : enabledMainSubtype) {
+            // Add main subtypes if they are not included in the user's dictionary
+            if (enabledUserDictionary != null && !enabledUserDictionary.contains(mainSubtype.getLocale())
+                    && !mainSubtype.getLocale().equals("zz")) {
+                sortedLanguages.add(mainSubtype.getLocale());
+            }
+            // Add secondary subtypes only if "Use system languages" setting is not enabled
+            if (!localeSystemOnly) {
+                for (Locale secondSubtype : Settings.getSecondaryLocales(prefs, mainSubtype.getLocale())) {
+                    // Add secondary subtypes if they are not included in the user's dictionary
+                    if (enabledUserDictionary != null && !enabledUserDictionary.contains(secondSubtype.toString())) {
+                        sortedLanguages.add(secondSubtype.toString());
+                    }
+                }
+            }
+        }
+
+        // Add the enabled system languages
+        for (Locale systemSubtype : enabledSystemLocale) {
+            sortedLanguages.add(systemSubtype.toString());
+        }
+
+
+        // Add preferences
+        if (sortedLanguages.isEmpty()) {
+            userDictGroup.addPreference(createUserDictionaryPreference(null));
+        } else {
+            for (String localeUserDictionary : sortedLanguages) {
+                userDictGroup.addPreference(createUserDictionaryPreference(localeUserDictionary));
+            }
+        }
+    }
+
+    /**
+     * Create a single User Dictionary Preference object, with its parameters set.
+     * @param localeString The locale for which this user dictionary is for.
+     * @return The corresponding preference.
+     */
+    protected Preference createUserDictionaryPreference(@Nullable final String localeString) {
+        final Preference newPref = new Preference(requireContext());
+        final Intent intent = new Intent(USER_DICTIONARY_SETTINGS_INTENT_ACTION);
+
+        if (null == localeString) {
+            newPref.setTitle(Locale.getDefault().getDisplayName());
+        } else {
+            if (localeString.isEmpty()) {
+                newPref.setTitle(getString(R.string.user_dict_settings_all_languages));
+            } else {
+                // Special treatment for the known languages with _zz and _ZZ types
+                if (localeString.endsWith("_zz") || localeString.endsWith("_ZZ")) {
+                    final int resId = requireContext().getResources().getIdentifier("subtype_"+localeString, "string", requireContext().getPackageName());
+                    newPref.setTitle(requireContext().getString(resId));
+                } else {
+                    newPref.setTitle(LocaleUtils.constructLocaleFromString(localeString).getDisplayName());
```

Comment:
Can't you just use `LocaleUtils.getLocaleDisplayNameInSystemLocale` here? Or does it give different results?
Btw duplicating code is rarely a good thing when you want the same string in both places.

Reply:
Done with commit f7f791c
As we don't have access to _Locale_ I updated `UserDictionarySettings.getLocaleDisplayName()` instead and applied it where necessary.